src/embed1p.jl

```diff
@@ -0,0 +1,88 @@
+
+
+import ACE, ACEbase
+
+import ACE: SList, i2val, val2i, get_spec, symbols, indexrange, get_index, degree 
+
+
+# -------------------------
+
+@doc raw"""
+`EmbedCat1pBasis` : defines an embedding of a categorical basis; 
+todo - document the details... 
+"""
+struct EmbedCat1pBasis{VSYM, ISYM, LEN, TCAT, T} <: Discrete1pBasis{T}
+   categories::SList{LEN, TCAT}
+   embed::Vector{Vector{T}}
+   label::String
+end
+
+_varsym(::EmbedCat1pBasis{VSYM, ISYM}) where {VSYM, ISYM} = VSYM
+_isym(::EmbedCat1pBasis{VSYM, ISYM}) where {VSYM, ISYM} = ISYM
+
+_val(X, B::EmbedCat1pBasis) = getproperty(X, _varsym(B))
+_idx(b, B::EmbedCat1pBasis) = getproperty(b, _isym(B))
+
+Base.length(B::EmbedCat1pBasis) = length( first(B.embed) )
+
+function EmbedCat1pBasis(categories::AbstractArray, embeddings::Dict; 
+              varsym::Symbol = :mu, idxsym::Symbol = :k, 
+              label = "EC$(idxsym)") 
+   list = SList(categories)
+   lenb = length( first(embeddings)[2] )
+   @assert all(lenb == length(x) for x in values(embeddings))
+   embed = [ embeddings[i2val(list, i)] for i = 1:length(categories) ]
+   return EmbedCat1pBasis(list, embed, varsym, idxsym, label)
+end
+
+EmbedCat1pBasis(categories::SList{LEN, TCAT}, embed::Vector{Vector{T}}, 
+                varsym::Symbol, isym::Symbol, label::String) where {LEN, TCAT, T} = 
+      EmbedCat1pBasis{varsym, isym, LEN, TCAT, T}(categories, embed, label)
+
+function ACE.evaluate!(A, basis::EmbedCat1pBasis, X::AbstractState)
+   key = _val(X, basis)
+   idx = val2i(basis.categories, key)
+   copy!(A, basis.embed[idx])
+   return A
+end
+
+ACE.valtype(::EmbedCat1pBasis{VSYM, ISYM, LEN, TCAT, T}, args...
+           ) where {VSYM, ISYM, LEN, TCAT, T} = T
+
+symbols(basis::EmbedCat1pBasis) = [ _isym(basis), ]
+
+indexrange(basis::EmbedCat1pBasis) = Dict( _isym(basis) => 1:length(basis) )
+
+isadmissible(b, basis::EmbedCat1pBasis) = (1 <= _idx(b, basis) <= length(basis))
+
+get_index(B::EmbedCat1pBasis, b) = _idx(b, B)
+
+degree(b, basis::EmbedCat1pBasis, args...) = 0
```

Comment:
Here is a key line: If we keep this `= 0` then it means that from the complexity perspective there is no difference between what we do here and the one-hot embedding. but if we were to assign a degree to each basis function then that would mean a path towards sparsification. E.g. the total degree basis would then select `S1` much more often than `S5`, say. I don't know what the canonical choice is here.